### PR TITLE
Proof-of-concept canvas Partymap implementation

### PR DESCRIPTION
**Pull request by mike-lvov**
This is a raw version of canvas partymap implementation. This is not ready for review. Creating a PR just so that you could play with it.

Basic technology used under the hood is Pixi.js (rendering engine, based on canvas).
I consider moving to somethings like Phaser just because it has physics engine as well. It would require less work for us to finish it then.

One issue that bothers me though is location exchange. Currently I use firebase store to update location. Would want to make it UDP-like rather than TCP-like. Maybe there are options in Firebase to emit actions rather then saving it in the db. 

A lot of work required for this to be in production!

### CODE REVIEW

**Comment by 0xdevalias** [comment]
> One issue that bothers me though is location exchange. Currently I use firebase store to update location. Would want to make it UDP-like rather than TCP-like. Maybe there are options in Firebase to emit actions rather then saving it in the db.

What is your issue with using firebase store in this way? And what would the 'flow' of a 'udp-like' implementation look like to you, at a high level?

Can you point out the relevant files/functions/etc where this currently exists to help narrow down where i'm looking for it?

**Comment by 0xdevalias** [comment]
Also, while this definitely adds a performance issue in that it causes a decent amount of processing initially to render the map/etc, I wouldn't say that it warrants a `P1 - Critical` label. I've changed it to `P2 - High` for now.. though in the scheme of our performance issues, I'm not even sure it is necessarily that drastic. I feel like some of the other performance issues in the app are far more critical/impactful.

File: src/components/templates/PartyMap/components/CanvasMap/charm.js [hunk 1]
```diff
@@ -0,0 +1,871 @@
+import * as PIXI from "pixi.js";
```
[the remark this answers is not in the document]
Comment (0xdevalias) [reply src/components/templates/PartyMap/components/CanvasMap/charm.js:1]: @mike-lvov I'm guessing this is a manually added copy/paste of https://github.com/kittykatattack/charm or similar?

File: src/components/templates/PartyMap/components/CanvasMap/CanvasMap.tsx [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+    //   firestore
+    //     .collection("venues")
+    //     .doc(venue.name)
+    //     .collection("onlinePlayers")
+    //     .doc(user.id)
+    //     .set({
+    //       x: player.x,
+    //       y: player.y,
+    //     });
+    // }, 200);
```
Comment (0xdevalias) [inline src/components/templates/PartyMap/components/CanvasMap/CanvasMap.tsx:242]: You don't need to have this in an interval, you could just have it in a `useEffect` or similar, with the player's position in the `deps`.

**Review by 0xdevalias — COMMENTED** [review]
I realise this is a WIP/not ready for a 'real' review, but in reading through the code to evaluate what your approach is, I figured I would make some running comments/notes as I went.

File: src/components/templates/PartyMap/components/CanvasMap/CanvasMap.tsx [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+  const firestore = useFirestore();
+  const on = useKeysListeners();
+
+  const divElementRef = useRef<HTMLDivElement>(null);
+  const [backgroundImageHeight, setBackgroundImageHeight] = useState("100%");
+  const appRef = useRef<{ app: PIXI.Application; charm: Charm } | null>(null);
+
+  useEffect(() => {
+    const divElement = divElementRef.current;
+    if (!divElement) return;
```
[the remark this answers is not in the document]
Comment (0xdevalias) [reply src/components/templates/PartyMap/components/CanvasMap/CanvasMap.tsx:47]: If we were going to explore using Phaser, I would want to make sure we weren't going to run into issues combining it with the rest of the React app, and that we do it in a neat/efficient way.

Haven't looked deeply into the topic, but a few things that popped up in a quick google:

- https://stackoverflow.com/questions/55667188/how-to-integrate-phaser-into-react
- https://tnodes.medium.com/integrating-react-and-phaser-3-tutorial-eb96717d4a9d
- https://medium.com/@larry.sassainsworth/the-wrong-way-to-integrate-phaser-with-react-d85e3b226cf9

**Comment by mike-lvov** [comment]
It is considered to be a proof-of-concept feature, but it seems that we have different/higher priorities now, so will close this PR for now @0xdevalias

**Pull request closed without merging**